File: 344.reverse-string.py
#
# @lc app=leetcode id=344 lang=python3
#
# [344] Reverse String
#

# @lc code=start
class Solution:
    def reverseHelper(self,s: List[str],left:int,right:int) -> None:
        if (left >= right):
            return
        s[left],s[right] = s[right],s[left]
        self.reverseHelper(s,left+1,right-1)
    def reverseString(self, s: List[str]) -> None:
        """
        Do not return anything, modify s in-place instead.
        """
        # #Recursion approach
        # self.reverseHelper(s,0,len(s)-1)

        #Two-Pointer
        left = 0
        right = len(s)-1
        while(left < right):
            s[left],s[right] = s[right],s[left]
            left += 1
            right -= 1

    
        
        

        
# @lc code=end

File: 344.reverse-string.c
/*
 * @lc app=leetcode id=344 lang=c
 *
 * [344] Reverse String
 */
#include <stdio.h>

// @lc code=start
void reverseHelper(char *s, int left, int right)
{
	// Base case : If the pointers meet or cross, the recursion stops.
	if (left >= right)
	{
		return;
	}
	// 1. Swap the characters at the current left and right boundaries.
	char temp = s[left];
	s[left] = s[right];
	s[right] = temp;
	// 2. Recursively call the function for the inner substring.
	reverseHelper(s, left + 1, right - 1);
}
void reverseString(char *s, int sSize)
{
	// // Recursion approach
	// if (s == NULL || sSize <= 1)
	// {
	// 	return;
	// }
	// reverseHelper(s, 0, sSize - 1);

	// Two-Pointer
	if (s == NULL || sSize <= 1)
	{
		return;
	}
	int left = 0;
	int right = sSize - 1;
	while (left < right)
	{
		char temp = s[left];
		s[left] = s[right];
		s[right] = temp;
		left++;
		right--;
	}
}
// @lc code=end
